packages/dotenv-webpack-plugin/src/index.js

```diff
@@ -0,0 +1,148 @@
+const dotenv = require("dotenv");
+const dotenvExpand = require("dotenv-expand");
+const { DefinePlugin } = require("webpack");
+const { validate } = require("schema-utils");
+const schema = require("./options.json");
+
+/** @typedef {import("./types").Config} Config */
+/** @typedef {import("schema-utils/declarations/validate").Schema} Schema */
+/** @typedef {import("webpack").Compiler} Compiler */
+
+class DotenvWebpackPlugin {
+  /**
+   * Dotenv Webpack Plugin
+   * @param {Config} options - Configuration options
+   */
+  constructor(options = {}) {
+    validate(/** @type {Schema} */ (schema), options || {}, {
+      name: "DotenvWebpackPlugin",
+      baseDataPath: "options",
+    });
+
+    const currentDirectory = process.cwd();
+
+    const {
+      envFiles = [
+        `${currentDirectory}/.env.example`, // loaded in all cases
+        `${currentDirectory}/.env`, // loaded in all cases
+        `${currentDirectory}/.env.local`, // loaded in all cases, ignored by git
+        `${currentDirectory}/.env.[mode]`, // only loaded in specified mode
+        `${currentDirectory}/.env.[mode].local`, // only loaded in specified mode, ignored by git
+      ],
+      prefixes = ["process.env.", "import.meta.env."],
+      envVarPrefix = "PUBLIC_",
+    } = options;
+
+    this.options = {
+      envFiles,
+      prefixes,
+      envVarPrefix,
+    };
+  }
+
+  /**
+   * Check if file exists
+   * @param {Compiler} compiler - Webpack compiler
+   * @param {string} environmentFile - Path to environment file
+   */
+  fileExists(compiler, environmentFile) {
+    return new Promise((resolve) => {
+      const fs = compiler.inputFileSystem;
+
+      fs.stat(environmentFile, (err) => {
+        // File does not exist
+        if (err) {
+          return resolve(false);
+        }
+
+        return resolve(environmentFile);
+      });
+    });
+  }
+
+  /**
+   * Read file from path and parse it
+   * @param {Compiler} compiler - Webpack compiler
+   * @param {string} environmentFile - Path to environment file
+   */
+  readFile(compiler, environmentFile) {
+    return new Promise((resolve, reject) => {
+      const envVariables = {};
+
+      const fs = compiler.inputFileSystem;
+
+      this.fileExists(compiler, environmentFile).then((exists) => {
+        if (!exists) {
+          return resolve();
+        }
+
+        fs.readFile(environmentFile, (err, environmentFileContents) => {
```

Comment:
Can we avoid `this.fileExists` and `fs.readFile`, because it is two fs calls, but we need only one, so if we can read file, let's read, otherwise let's log without `reject`

Reply:
done. only using `readFile` now